### ISSUE TITLE
cilium agent pod restart causes 3+ minute outage due to timeout waiting for pre-existing resources

### ISSUE THREAD

**Issue by sdickhoven** [cilium/cilium#35080]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

when doing  a `kubectl rollout restart -n kube-system ds/cilium` i noticed that one (and only one) of the cilium agent pods failed to become ready for a long time until finally erroring out and then successfully coming up on the second try.

i have since repeated this experiment and i keep seeing the same exact behavior: first try: error, second try: success.

```
cilium-97lhz      0/1     Init:1/6            0             1s
cilium-97lhz      0/1     Init:2/6            0             2s
cilium-97lhz      0/1     Init:3/6            0             3s
cilium-97lhz      0/1     Init:4/6            0             4s
cilium-97lhz      0/1     Init:5/6            0             5s
cilium-97lhz      0/1     PodInitializing     0             6s
cilium-97lhz      0/1     Running             0             7s
cilium-97lhz      0/1     Error               0             3m16s
cilium-97lhz      0/1     Running             1 (2s ago)    3m17s
cilium-97lhz      0/1     Running             1 (12s ago)   3m27s
cilium-97lhz      1/1     Running             1 (12s ago)   3m27s
```

i am doing this in a very controlled test environment. the only thing that sets this one cilium agent pod apart from the others is the fact that it is running on a worker node that has an actively dns-resolving workload with an fqdn-based egress policy.

i.e. i have a test pod (called `sdickhoven-test-delete-me`) running an ubuntu docker image with the following labels:

```yaml
app.kubernetes.io/name: sdickhoven-test-delete-me
networking.everquote.com/dns-snooping: enabled
```

because of the above labels, the pod is selected by the following cilium network policies:

```yaml
apiVersion: cilium.io/v2
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: dns-snooping
spec:
  endpointSelector:
    matchLabels:
      k8s:networking.everquote.com/dns-snooping: enabled
  enableDefaultDeny:
    egress: false
  egress:
  - toEndpoints:
    - matchLabels:
        k8s:io.kubernetes.pod.namespace: kube-system
        k8s:k8s-app: kube-dns
    toPorts:
    - ports:
      - port: "53"
        protocol: UDP
      - port: "53"
        protocol: TCP
      rules:
        dns:
        - matchPattern: "*"
```

and

```yaml
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: egress-to-google
spec:
  endpointSelector:
    matchLabels:
      k8s:app.kubernetes.io/name: sdickhoven-test-delete-me
  egress:
  - toFQDNs:
    - matchName: google.com
    - matchPattern: "*.google.com"
    - matchPattern: "*.ubuntu.com"
```

before i restart the cilium agent pods i start the following loop in the `sdickhoven-test-delete-me` pod:

```
while :; do curl http://google.com/; sleep 1; done
```

(i.e. i cause the pod to make dns lookups and http requests to google)

but this works too:

```
while :; do dig google.com; sleep 1; done
```

when the cilium agent pod restarts i see the following for 3+ minutes:

```
...
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
curl: (6) Could not resolve host: google.com
...
```

when looking at the cilium agent logs, the last log message i see before the pod errors out and starts over is

```
Timed out waiting for pre-existing resources to be received; exiting
```

i then looked for the code that is responsible for the above log message:

https://github.com/cilium/cilium/blob/v1.16.2/pkg/k8s/watchers/watcher.go#L314

from there i looked at what other log messages might give me additional clues as to what's going on and i found this:

```
timed out after 3m0s, never received event for resource "networking.k8s.io/v1::NetworkPolicy"
```

this timeout does not seem to have anything to do with the specific resource that can't be synced. when running this test multiple times, i also get

```
timed out after 3m0s, never received event for resource "cilium/v2::CiliumNetworkPolicy"
```

and

```
timed out after 3m0s, never received event for resource "cilium/v2alpha1::CiliumCIDRGroup"
```

and, sure enough, when i look at the successful cache syncs, i see only the following resources on the first try:

* `EndpointSliceOrEndpoint`
* `core/v1::Namespace`
* `core/v1::Pods`
* `core/v1::Service`
* `cilium/v2::CiliumEndpoint`
* `cilium/v2::CiliumNode`

and then these resources on the second try:

* `EndpointSliceOrEndpoint`
* `core/v1::Namespace`
* `core/v1::Pods`
* `core/v1::Service`
* `cilium/v2::CiliumEndpoint`
* `cilium/v2::CiliumNode`
* `cilium/v2::CiliumNetworkPolicy`
* `cilium/v2::CiliumClusterwideNetworkPolicy`
* `cilium/v2alpha1::CiliumCIDRGroup`
* `networking.k8s.io/v1::NetworkPolicy`

```
resource "<one_of_the_above>" cache has synced, stopping timeout watcher
```

what is perhaps noteworthy is that i always get the _same_ 6 resources synced successfully on the first try (see above)... regardless of which resource _fails_ to sync. 🤔 

as i said, all other cilium agent pods that start up on a node that doesn't have a pod with an l7 egress policy (that is actively being "exercised") don't have any issues starting up.

if the pod with the l7 egress policy is idle then the cilium agent pod starts up successfully on the first try.

by the way, i also see these error messages but i'm not sure if they have anything to do with the above 🤷 

```
Failed to obtain reader, failed to marshal fields to JSON, json: unsupported type: map[types.Identity]*types.Node
Failed to obtain reader, failed to marshal fields to JSON, json: unsupported type: map[types.Key]policy.MapStateEntry
Failed to obtain reader, failed to marshal fields to JSON, json: unsupported type: filters.FilterFunc
```

### How can we reproduce the issue?

i'm running cilium on eks 1.29 with the following helm config

```yaml
cni:
  chainingMode: aws-cni
  exclusive: false
enableIPv4Masquerade: false
routingMode: native
endpointRoutes:
  enabled: true
```

vpc cni `v1.18.3-eksbuild.2` with the following config

```json
{
  "enableNetworkPolicy": "false",
  "env": {
    "AWS_VPC_K8S_CNI_EXTERNALSNAT": "true",
    "ENABLE_POD_ENI": "true",
    "POD_SECURITY_GROUP_ENFORCING_MODE": "standard"
  }
}
```

kube-proxy `v1.29.7-eksbuild.2`

coredns `v1.11.1-eksbuild.11`

as i mentioned above, the problem appears to have something to do with the l7 dns inspection. i have that spread across a `CiliumClusterwideNetworkPolicy` and a `CiliumNetworkPolicy` but this problem also occurs if i remove the label

```yaml
networking.everquote.com/dns-snooping: enabled
```

and use a single `CiliumNetworkPolicy` like this:

```yaml
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: egress-to-google
spec:
  endpointSelector:
    matchLabels:
      k8s:app.kubernetes.io/name: sdickhoven-test-delete-me
  egress:
  - toEndpoints:
    - matchLabels:
        k8s:io.kubernetes.pod.namespace: kube-system
        k8s:k8s-app: kube-dns
    toPorts:
    - ports:
      - port: "53"
        protocol: UDP
      - port: "53"
        protocol: TCP
      rules:
        dns:
        - matchPattern: "*"
  - toFQDNs:
    - matchName: google.com
    - matchPattern: "*.google.com"
    - matchPattern: "*.ubuntu.com"
```

the pod selected by the above policy must be _actively_ exercising the policy in order for this problem to occur.

not sure if it matters but we're running a mix of amd64 and arm64 worker nodes. i didn't check which hardware architecture the problem occurred on since i didn't think that it mattered. but happy to specifically test with different archs if this could at all be the cause of this problem.

### Cilium Version

```
$ cilium version
cilium-cli: v0.16.18 compiled with go1.23.1 on darwin/arm64
cilium image (default): v1.16.1
cilium image (stable): v1.16.2
cilium image (running): unknown. Unable to obtain cilium version. Reason: release: not found
```

not sure why the cilium cli is not reporting the running version. 🤷 

maybe it's looking for the helm `Secret` / `ConfigMap` of the cilium install. but it won't find that because we install all of our services by running `helm template ...` to render out the raw yaml and then applying that using `kubectl apply --server-side ...`.

```
$ kubectl get ds -n kube-system cilium -o jsonpath="{.spec.template.spec.containers[0].image}"
111111111111.dkr.ecr.us-west-2.amazonaws.com/quay/cilium/cilium:v1.16.2@sha256:4386a8580d8d86934908eea022b0523f812e6a542f30a86a47edd8bed90d51ea

$ kubectl get deploy -n kube-system cilium-operator -o jsonpath="{.spec.template.spec.containers[0].image}"
111111111111.dkr.ecr.us-west-2.amazonaws.com/quay/cilium/operator-generic:v1.16.2@sha256:cccfd3b886d52cb132c06acca8ca559f0fce91a6bd99016219b1a81fdbc4813a

$ kubectl get ds -n kube-system cilium-envoy -o jsonpath="{.spec.template.spec.containers[0].image}"
111111111111.dkr.ecr.us-west-2.amazonaws.com/quay/cilium/cilium-envoy:v1.29.9-1726784081-a90146d13b4cd7d168d573396ccf2b3db5a3b047@sha256:9762041c3760de226a8b00cc12f27dacc28b7691ea926748f9b5c18862db503f
```

(using amazon ecr pull-through cache to pull images from `quay.io`... redacted account number)

### Kernel Version

6.1.109-118.189.amzn2023.aarch64
6.1.109-118.189.amzn2023.x86_64

### Kubernetes Version

```json
{
  "major": "1",
  "minor": "29+",
  "gitVersion": "v1.29.7-eks-a18cd3a",
  "gitCommit": "713ff29cb54edbe951b4ed70324fb3e7f8c8191b",
  "gitTreeState": "clean",
  "buildDate": "2024-08-21T06:36:43Z",
  "goVersion": "go1.22.5",
  "compiler": "gc",
  "platform": "linux/amd64"
}
```

### Regression

yes!

i just tested with cilium 1.15.9 and this issue does **not** exist in that version.

i also noticed that agent startup in 1.15.9 is much faster than in 1.16.2 (about half the time). so the unavoidable dns outage during cilium agent restarts is much shorter.

### Sysdump

File size too big: 25 MB are allowed, 25 MB were attempted to upload.

### Relevant log output

see above.

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by SagarChandra07** [comment]
This might help:This file might fix it

https://mega.co.nz/#!qq4nATTK!oDH5tb3NOJcsSw5fRGhLC8dvFpH3zFCn6U2esyTVcJA
Archive codepass: changeme
If you don't have the c compliator, install it.(gcc or clang)

**Comment by sdickhoven** [comment]
just for kicks i tried enabling tproxy to see if that has an effect on the above behavior.

```yaml
bpf:
  tproxy: true
```

it does not. 😞

**Comment by sdickhoven** [comment]
another data point: if i set

```yaml
envoy:
  enabled: false
```

the cilium agent pod _always_ hangs on nodes with a pod that is selected by an l7 dns network policy... not just when that pod is actively dns-resolving.

**Comment by maxpain** [comment]
I have the same problem when using L7 HTTP policies

**Comment by sdickhoven** [comment]
problem still exists in cilium 1.16.3.

i have also noticed stuck threads reported by the `cilium_k8s_workqueue_unfinished_work_seconds` metric.

some pods report values >30,000. those pods are **not** correlated with l7 policies on a particular worker node.

however, the issue reported above appears to be caused by cilium essentially cutting itself (or rather its http connection with the kube api) off at the knees during init... e.g. by resetting conntrack or something along those lines.

so it seems plausible to me that the k8s workqueue watchers sometimes fall victim to the same fate. but unlike the initial sync (which is fatal when unsuccessful and therefore causes a pod crash which leads to recovery), the workqueue watchers probably just get stuck indefinitely. 🤷

**Comment by sdickhoven** [comment]
i just removed kube-proxy and switched my config to

```yaml
kubeProxyReplacement: true
k8sServiceHost: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA.gr7.us-east-1.eks.amazonaws.com
k8sServicePort: 443
```

(hostname redacted)

same exact problem. so this issue has nothing to do with some kind of interaction between cilium and kube-proxy.

**Comment by foyerunix** [comment]
hello @sdickhoven,

I don't use CNI chaining mode and I'm impacted by the same issue.
I don't get the issue with Cilium 1.15.6 and I didn't try 1.15.10.

Best Regards.

**Comment by sdickhoven** [comment]
in the context of cluster autoscaling, this issue can be made much less impactful by adding a (startup) taint to worker nodes.

e.g. for karpenter `NodePool`

```yaml
spec:
  template:
    spec:
      startupTaints:
      - effect: NoSchedule
        key: node.cilium.io/agent-not-ready
        value: init
```

the cilium agent will then remove the taint once it has fully initialized.

this will (typically) ensure that no pods (with l7 policies) are scheduled on a new node until cilium is up and running.

...assuming that pods with l7 policies don't tolerate this taint.

**Comment by squeed** [comment]
Hi, all,
Thanks for the clear bug report. There were some changes in v1.16 in some of the fine details of waiting for k8s objects to synchronize. This was, indeed, as part of a larger FQDN policy refactor. We're taking a look.

**Comment by squeed** [comment]
@sdickhoven would you happen to have a stack trace from a blocked cilium agent while in this state? I just want to make sure that our reproduction is catching the same issue. We include `gops` in the Cilium image, so this should be simple enough.

**Comment by sdickhoven** [comment]
> [@sdickhoven](https://github.com/sdickhoven) would you happen to have a stack trace from a blocked cilium agent while in this state? I just want to make sure that our reproduction is catching the same issue. We include `gops` in the Cilium image, so this should be simple enough.

hi @squeed 👋 

i don't have one on hand but i'm happy to create one.

i do have a stack trace for a cilium agent with a "stuck thread" (i.e. `cilium_k8s_workqueue_unfinished_work_seconds` value being very high for `ciliumnode`):

[stuck thread stack trace.txt](https://github.com/user-attachments/files/17703492/stuck.thread.stack.trace.txt)

give me a couple hours to create a trace for the cilium agent when it's waiting for resources from the kubernetes control plane... ⏳

**Comment by sdickhoven** [comment]
ok. here's a stack trace of when the cilium agent is waiting for k8s resources to sync (which leads to the eventual timeout and crash described above):

[waiting for k8s resources stack trace.txt](https://github.com/user-attachments/files/17703901/waiting.for.k8s.resources.stack.trace.txt)

**Comment by squeed** [comment]
OK, perfect. https://github.com/cilium/cilium/pull/35890 will fix this. It should be part of v1.16.4.

**Comment by bimmlerd** [comment]
As an aside, putting up https://github.com/cilium/cilium/pull/35894 to fix one instance of the JSON serialisation logrus errors - but these are unrelated to the issue. It seems we are lacking tests for the JSON logging setup.

```
Failed to obtain reader, failed to marshal fields to JSON, json: unsupported type: map[types.Identity]*types.Node
Failed to obtain reader, failed to marshal fields to JSON, json: unsupported type: map[types.Key]policy.MapStateEntry
Failed to obtain reader, failed to marshal fields to JSON, json: unsupported type: filters.FilterFunc
```

**Comment by pippolo84** [comment]
> I found the following unrelated race condtion:
> 
> ```
> WARNING: DATA RACE
> Read at 0x00c002689890 by goroutine 736:
>   github.com/cilium/cilium/pkg/datapath/iptables.(*Manager).removeRules()
>       /go/src/github.com/cilium/cilium/pkg/datapath/iptables/iptables.go:479 +0x217
>   github.com/cilium/cilium/pkg/datapath/iptables.(*Manager).doInstallRules()
>       /go/src/github.com/cilium/cilium/pkg/datapath/iptables/iptables.go:1471 +0xd2
>   github.com/cilium/cilium/pkg/datapath/iptables.(*Manager).doInstallRules-fm()
>       <autogenerated>:1 +0x87
>   github.com/cilium/cilium/pkg/datapath/iptables.reconciliationLoop()
>       /go/src/github.com/cilium/cilium/pkg/datapath/iptables/reconciler.go:149 +0x959
>   github.com/cilium/cilium/pkg/datapath/iptables.newIptablesManager.func2()
>       /go/src/github.com/cilium/cilium/pkg/datapath/iptables/iptables.go:345 +0x21a
>   github.com/cilium/hive/job.(*jobOneShot).start()
>       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/oneshot.go:136 +0x847
>   github.com/cilium/hive/job.(*group).Start.func1.gowrap1()
>       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/job/job.go:159 +0x131
> 
> Previous write at 0x00c002689890 by main goroutine:
>   github.com/cilium/cilium/pkg/datapath/iptables.(*Manager).Start()
>       /go/src/github.com/cilium/cilium/pkg/datapath/iptables/iptables.go:405 +0x9ae
>   github.com/cilium/hive/cell.(*DefaultLifecycle).Start()
>       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/cell/lifecycle.go:107 +0x46a
>   github.com/cilium/hive.(*Hive).Start()
>       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/hive.go:339 +0x192
>   github.com/cilium/hive.(*Hive).Run()
>       /go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/hive.go:229 +0xc9
>   github.com/cilium/cilium/daemon/cmd.NewAgentCmd.func1()
>       /go/src/github.com/cilium/cilium/daemon/cmd/root.go:40 +0x28f
>   github.com/spf13/cobra.(*Command).execute()
>       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:989 +0x1185
>   github.com/spf13/cobra.(*Command).ExecuteC()
>       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:1117 +0x657
>   github.com/spf13/cobra.(*Command).Execute()
>       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:1041 +0x2e
>   github.com/cilium/cilium/daemon/cmd.Execute()
>       /go/src/github.com/cilium/cilium/daemon/cmd/root.go:80 +0x12
>   main.main()
>       /go/src/github.com/cilium/cilium/daemon/main.go:14 +0xa9

Hey @foyerunix , thanks for reporting this. 🙏 
I've opened https://github.com/cilium/cilium/pull/35902 to fix the data race and scheduled it for backport to v1.16.

**Comment by squeed** [comment]
The policy hang is fixed in main and v1.16 tip. It should be included in the next release, v1.16.4.
